runtime/compiler/codegen/J9TreeEvaluator.cpp

```diff
@@ -35,6 +35,328 @@
+   {
+   TR_ASSERT_FATAL(!node->getOpCode().isIndirect(), "A J9JITWatchedStaticFieldData structure can only be initialized for static nodes: " POINTER_PRINTF_FORMAT, node);
+   // Prepare data block.
+   staticFieldDataBlock->method = owningMethod;
+   staticFieldDataBlock->location = bcIndex;
+   if (isUnresolved)
+      {
+      // If unresolved at compile time, we initialize the fieldAddress to -1. We will eventually call generateFillInDataBlockSequenceForUnresolvedField
+      // to generate instructions that will resolve and correctly populate this field at runtime.
+      staticFieldDataBlock->fieldAddress = reinterpret_cast<void *>(-1);
+      }
+   else
+      {
+      TR::SymbolReference *symRef = node->getSymbolReference();
+      J9Class *fieldClass = reinterpret_cast<J9Class *>(symRef->getOwningMethod(comp)->getDeclaringClassFromFieldOrStatic(comp, symRef->getCPIndex()));
+      TR_ASSERT_FATAL(fieldClass, "Valid field class is expected for resolved field reference " POINTER_PRINTF_FORMAT, node);
+      staticFieldDataBlock->fieldAddress = static_cast<void *>(node->getSymbolReference()->getSymbol()->getStaticSymbol()->getStaticAddress());
+      staticFieldDataBlock->fieldClass = fieldClass;
+      }
+   }
+
+static TR::Snippet *prepareDataSnippetForFieldWatch(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT_FATAL(J9ClassHasWatchedFields >= std::numeric_limits<uint16_t>::min() && J9ClassHasWatchedFields <= std::numeric_limits<uint16_t>::max(), "Expecting value of J9ClassHasWatchedFields to be within 16 bits. Currently it's %d(%p).", J9ClassHasWatchedFields, J9ClassHasWatchedFields);
+
+   J9Method *owningMethod = reinterpret_cast<J9Method *>(node->getOwningMethod());
+   int32_t bcIndex = node->getByteCodeInfo().getByteCodeIndex();
+   bool isUnresolved = node->getSymbolReference()->isUnresolved();
+   bool isStatic = node->getSymbolReference()->getSymbol()->isStatic();
+   TR::Snippet *dataSnippet;
+   if (isStatic)
+      {
+      // For wrtbar nodes the j9classReg is the grandchild of the second child and the evaluation of the second child is optimized
+      // to avoid explicitly storing the j9class into a register. Evaluate the j9class node to store the j9class into j9classReg.
+      if (node->getOpCode().isWrtBar() && isUnresolved)
+         {
+         cg->evaluate(TR::TreeEvaluator::getDirectWrtbarJ9ClassNode(node));
+         }
+      // Create structure and populate it with data required by the VM Helper.
+      static J9JITWatchedStaticFieldData staticFieldDataBlock;
+      prepareStaticFieldDataForFieldWatch(node, owningMethod, bcIndex, cg->comp(), &staticFieldDataBlock, isUnresolved);
+      // Create data snippet to store the structure.
+      dataSnippet = TR::TreeEvaluator::createDataSnippetForFieldWatch(cg, &staticFieldDataBlock, node, isStatic);
```

Comment:
@dsouzai can you please review this function? It gets the `J9Class*` in `prepareStaticFieldDataForFieldWatch` via `getDeclaringClassFromFieldOrStatic`. Will this be relocated properly in AOT? I don't see any relocation here so I'm guessing this will fail, however I'd like an expert opinion.

Reply:
Because `createDataSnippetForFieldWatch` is calling `CreateConstant`, it looks like in `TR::S390ConstantDataSnippet::emitSnippetBody()` we do create relocations as long as the relotype is set. So I guess what needs to be added here is something like 
```
dataSnippet->setReloType(TR_ClassAddress)
```

Reply:
But how does the relocation know where inside the snippet the relocation is supposed to happen?

Reply:
However, I think this will only work for AOT w/ SVM. I don't think there's a way to get it to work for non-SVM.

Reply:
Just to be clear the `J9Class*` is not in a snippet here directly, it's in a filed of some `struct` which is being stored into a snippet:

```
typedef struct J9JITWatchedStaticFieldData {
	J9Method *method;		/* Currently executing method */
	UDATA location;			/* Bytecode PC index */
	void *fieldAddress;		/* Address of static field storage */
	J9Class *fieldClass;	/* Declaring class of static field */
} J9JITWatchedStaticFieldData;
```
So if I'm understanding this correctly we need to check `needClassAndMethodPointerRelocations` and then create proper relocations for several things in here. Correct?

Reply:
Oh I see; I'd say this thing needs its own relokind. There's too many different things to relocate and would make the code messy.

Reply:
Specifically, the new reloKind is going to have to allocate a `J9JITWatchedStaticFieldData`, populate it, and then patch the snippet with the address of the `J9JITWatchedStaticFieldData` struct.

Reply:
Btw, why is it a static in the compiler code??

```
static J9JITWatchedStaticFieldData staticFieldDataBlock;
```

Reply:
> Btw, why is it a static in the compiler code??

Good catch. This is a bug.

w.r.t. it's own relocation kind I'll delegate that to @dchopra001 as I'm not sure how `location` and `fieldAddress` are used.

Reply:
The structs are no longer `static`: 90bc28a